cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/Roller.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright 2018, Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.cluster.operator.resource;
+
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.api.model.apps.StatefulSet;
+import io.strimzi.operator.common.operator.resource.PodOperator;
+import io.vertx.core.Future;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * <p>Abstracted algorithm for rolling the pods of a StatefulSet.</p>
+ *
+ * <p>The algorithm works as follows:</p>
+ * <pre>
+ * 1. Get a (mutable) context to be passed around during the
+ *    roll (by applying {@link #context(StatefulSet, Secret, Secret)}.
+ *    This context is implementation dependent.
+ * 2. {@link Context#sort(Context, Predicate)} the context. This could,
+ *    for example, influence the order in which pods and rolled and when pods are rolled.
+ * 3. While {@link Context#next()} returns a non-null value:
+ *      a. Precondition
+ *      b. Restart
+ *      c. Postcondition
+ *      d. Continue from step 2.
+ * </pre>
+ *
+ * @param <P> The class representing pods.
+ *           E.g. could be {@code Pod}, or {@code String} (pod name), or {@code Integer} (pod number).
+ * @param <C> A context for the roll, passed along during the rolling restart.
+ */
+abstract class Roller<P, C extends Roller.Context<P>> {
+
+    private final Logger log = LogManager.getLogger(getClass());
+
+    private static final String NO_UID = "NULL";
+    protected final long operationTimeoutMs;
+    protected final PodOperator podOperations;
+
+    public Roller(long operationTimeoutMs, PodOperator podOperations) {
+        this.operationTimeoutMs = operationTimeoutMs;
+        this.podOperations = podOperations;
+    }
+
+    public interface Context<P> {
+
+        /** @return the next pod to be rolled, or null if there are no more pods to roll */
+        P next();
+
+        String toString();
+    }
+
+    /**
+     * Get the rolling context for restarting the pods in the given StatefulSet.
+     */
+    protected abstract Future<C> context(StatefulSet ss,
+                               Secret clusterCaCertSecret, Secret coKeySecret);
+
+    /**
+     * Sort the context, potentially changing the next pod to be rolled, etc.,
+     * returning a Future that completes when the sorting is complete.
+     */
+    protected abstract Future<C> sort(C context, Predicate<Pod> podRestart);
+
+    /**
+     * Returns a future which completes when the given pod is ready to be rolled
+     */
+    protected abstract Future<Void> beforeRestart(Pod pod);
```

Comment:
Could it be useful returning a `Boolean` if the pre-condition in the `beforeRestart` somehow wants to avoid restarting?

Reply:
That's what the poorly-named `podRestart` is for. And they're necessarily distinct concerns. It's the KAO which "knows" whether a pod needs to be restarted for some reason, but it's the KSO or ZSO which would know what precondition/barrier was necessary before a pod could be safely restarted. In actual fact neither the Kafka nor ZK rolling actually needs a precondition with the current code, but we do use the post condition, and the same comments apply to that.